iroh-bytes/src/store/fs.rs

```diff
@@ -1643,10 +1646,40 @@ impl ActorState {
                                     tx.send(Ok(())).ok();
                                 }
                                 Err(e) => {
-                                    drop(temp_tag);
-                                    tx.send(Err(e.into())).ok();
+                                    const ERR_CROSS: i32 = 18;
+                                    if e.raw_os_error() == Some(ERR_CROSS) {
+                                        // Cross device renaming failed, copy instead
+                                        match std::fs::copy(&path, &target) {
+                                            Ok(_) => {
+                                                let entry = EntryState::Complete {
+                                                    data_location: DataLocation::External(
+                                                        vec![target],
+                                                        size,
+                                                    ),
+                                                    outboard_location,
+                                                };
+                                                // Delete old file
+                                                if let Err(err) = std::fs::remove_file(&path) {
```

Comment:
You are not supposed to actually delete stuff here. If the transaction gets rejected the store would be in an inconsistent state. The entry would still point to internal data, but that internal data would no longer be there.

The proper way to delete things is to just _mark_ them for deletion by adding this file to the delete_after_commit set.

Unfortunately this is a bit convoluted, but keeping a file system and a transactional DB in sync is a PITA.

                                                 ```
-                                                // Delete old file
-                                                if let Err(err) = std::fs::remove_file(&path) {
-                                                    warn!("failed to delete {} after exporting it: {:?}", path.display(), err);
-                                                }
-
                                                 drop(guard);
                                                 tables.blobs.insert(temp_tag.hash(), entry)?;
+                                                tables
+                                                    .delete_after_commit
+                                                    .insert(*temp_tag.hash(), [BaoFilePart::Data]);
                                                 drop(temp_tag);
                                                 ```

Reply:
thanks, fixed